pytorch_lightning/callbacks/base.py

```diff
@@ -267,6 +267,10 @@ def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningM
         """Called when the training is interrupted by ``KeyboardInterrupt``."""
         pass
 
+    def on_exception(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", exception: BaseException) -> None:
+        """Called when any trainer execution is interrupted by an exception."""
```

Comment:
do we keep it for any exception regardless of train/valid/test or during init?

Reply:
not sure I fully understand the question. This will be called for the main trainer function entry points (fit, validate, test, predict) - Currently it won't be called for trainer init. However, we could extend error handling to init as well, but that would be in a separate PR

Reply:
Is it an exception only for the train loop, or also the validation loop?
if so how callback can differentiate from which loop the exception came? so add it as a callback argument?

Reply:
yes it works for train loop and validation loop. you're right currently the callback can't differentiate from which loop it came. but i'm not sure if that is a requested feature. I'd suggest we merge the simple version for now, and if users request that additional feature we can add it. wdyt? do others have thoughts on this? @tchaton @ananthsub

Reply:
great point @Borda !

@daniellepintz i think we should address this in the issue to ensure the design fully covers:
- When exactly is this callback called during the exception handling
- At that point, what metadata is available on the trainer object? This would inform what metadata ought to be passed through the hook
- Should end users be able to swallow the exceptions or only be able to re-raise a different exception?
- The relationship between this and https://github.com/PyTorchLightning/pytorch-lightning/issues/8895

Reply:
oh, I just realized the user can access the trainer state with `trainer.state.stage`. We pass the whole trainer object through the hook so the user has access to all the metadata.

regarding when this callback is called during exception handling, I think the current placement makes sense but lmk if you think it should be moved. and I think end users should be able to swallow the exceptions as discussed with @tchaton below, but I ran into an issue implementing it as I mention below